### ISSUE TITLE
[BUG] Behaviour of gradient transforms with `argnum=[]`

### ISSUE THREAD

**Issue by dwierichs** [PennyLaneAI/pennylane#4213]
### Expected behavior

Users that explicitly provide an empty container to the `argnum` (or `argnums` for JAX) keyword argument likely expect that no differentiation takes place, i.e. that they obtain an empty derivative.

### Actual behavior

The gradient transforms do various things for `argnum=[]`, but do not return an empty derivative. This is not a drastic problem for users that access the gradients module indirectly, by specifying the `diff_method` of a `QNode` and using ML-interface functions like `jax.jacobian`. When using the gradients module directly, however, the output sometimes is incorrect, sometimes an error is raised.

### Source code

```shell
dev = qml.device('default.qubit', wires=2)

@qml.qnode(dev)
def gate_node(x, y):
    qml.RX(x, 0)
    qml.RY(y, 0)
    return qml.expval(qml.PauliZ(0))

x = pnp.array(0.4, requires_grad=True)
y = pnp.array(0.1, requires_grad=True)

>>> qml.gradients.finite_diff(gate_node, argnum=[])(x, y) # Wrong
(tensor(0., requires_grad=True), tensor(0., requires_grad=True))
>>> qml.gradients.spsa_grad(gate_node, argnum=[])(x, y) # Wrong
(tensor(0., requires_grad=True), tensor(0., requires_grad=True))
>>> qml.gradients.param_shift(gate_node, argnum=[])(x, y)
raises IndexError
>>> qml.gradients.hadamard_grad(gate_node, argnum=[])(x, y) # Correct
(tensor(-0.38747287, requires_grad=True), tensor(-0.09195267, requires_grad=True))

H = qml.dot([qml.pulse.constant] * 2, [qml.PauliX(0), qml.PauliY(0)])
x, y = jnp.array([0.5, 0.2])

@qml.qnode(dev)
def pulse_node(x, y):
    qml.evolve(H)([x, y], t=0.2)
    return qml.expval(qml.PauliZ(0))

>>> qml.gradients.stoch_pulse_grad(pulse_node, argnum=[])(x, y)
raises UnboundLocalError
```

### System information

```shell
PL dev
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.
